File: array/0053_maximum_subarray/0053_maximum_subarray.go
package main

import (
	"fmt"
)

func maxSubArray(nums []int) int {

	if len(nums) == 1 {
		return nums[0]
	}
	maxSum, sumSoFar := nums[0], nums[0]
	for i := 1; i < len(nums); i++ {
		if nums[i] > sumSoFar+nums[i] {
			sumSoFar = nums[i]
		} else {
			sumSoFar += nums[i]
		}
		if sumSoFar > maxSum {
			maxSum = sumSoFar
		}
	}

	return maxSum
}
func main() {
	nums := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	fmt.Println(maxSubArray(nums))
}

File: array/0053_maximum_subarray/0053_maximum_subarray.py
# -*- coding: utf-8 -*-


class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if max(nums) < 0:
            return max(nums)
        
        global_max, local_max = float('-inf'), 0
        for num in nums:
            local_max = max(0, local_max + num)
            global_max = max(local_max, global_max)
        return global_max



nums = [-2,1,-3,4,-1,2,1,-5,4]
print(Solution().maxSubArray(nums))